### ISSUE TITLE
[Feature][EPDDisagg] Support distributed cache as implementation for EC connector

### ISSUE THREAD

**Issue by wuhang2014** [JiusiServe/vllm#61]
**1、 Mooncake store EC connector的目标：**
Encoder实例与PD实例分离后，Encoder计算出的encoder需要传输到PD实例进行复用。Mooncake store 是一个分布式缓存池，Encoder实例在计算完encoder后，会将encoder保存到mooncake store中； PD实例在计算kvc前，加载需要的encoder cache。

方案设计参考郑博提交的PR https://github.com/vllm-project/vllm/pull/25233:
EC connector的函数实现参考 vllm/distributed/ec_transfer/ec_connector/base.py

- 使用：

Encoder (producer) example:
```
vllm serve <model> \
  --ec-transfer-config '{
    "ec_connector": "ECMooncakeStorageConnector",
    "ec_role": "ec_producer",
    "ec_connector_extra_config": {
    "local_hostname": "localhost", #本机节点名，多机部署时使用ip
    "metadata_server": "http://localhost:8080/metadata", #元数据服务地址，端口号在启动mooncake时配置，多机部署使用ip
    "global_segment_size": 32212254720, #全局预留空间大小，30G
    "local_buffer_size": 1073741824, #本地缓冲区大小, 1G
    "protocol": "tcp", # 通信协议
    "device_name": "", #设备， 不填写时自动查询
    "master_server_address": "localhost:50051", #master服务地址，端口号在启动mooncake时配置，多机部署使用ip
    "replica_num": 1, #存储副本数
    "fast_transfer": true, #是否开启零拷贝传输开关
    "fast_transfer_buffer_size": 1 #零拷贝传输预留缓冲大小
  }'

```

PD (consumer) example:
```
vllm serve <model> \
  --ec-transfer-config '{
    "ec_connector": "ECSharedStorageConnector",
    "ec_role": "ec_consumer",
    "ec_connector_extra_config": {
    "local_hostname": "localhost",
    "metadata_server": "http://localhost:8080/metadata",
    "global_segment_size": 32212254720,
    "local_buffer_size": 1073741824,
    "protocol": "tcp",
    "device_name": "",
    "master_server_address": "localhost:50051",
    "replica_num": 1,
    "fast_transfer": true,
    "fast_transfer_buffer_size": 1
  }'
```

- connector各函数的调用时机：
1）encoder实例查询多模态数据是否有缓存命中(has_cache):
vllm/v1/core/sched/scheduler.py
```
    if request.has_encoder_inputs:
        (encoder_inputs_to_schedule, num_new_tokens,
         new_encoder_compute_budget, external_load_encoder_input
         ) = self._try_schedule_encoder_inputs(
             request, request.num_computed_tokens, num_new_tokens,
             encoder_compute_budget)
```
请求中命中缓存的多模态数据项，如果命中缓存，则不需要重复计算encoder。
会先在显存中查询是否存在encoder cache，显存中不存在，从connector的存储后端中查询。

2）encoder实例计算以及保存encoder(save_caches)的代码：
vllm/v1/worker/gpu_model_runner.py
```
    if has_ec_transfer() and get_ec_transfer().is_producer:
            with self.maybe_get_ec_connector_output(
                    scheduler_output,
                    encoder_cache=self.encoder_cache,
            ) as ec_connector_output:
                self._execute_mm_encoder(scheduler_output)
                return make_empty_encoder_model_runner_output(scheduler_output)
```
Encode实例在计算并保存encode后直接返回，不进行kv计算。

3）PD实例查询encoder(has_cache)是否有缓存命中:
vllm/v1/core/sched/scheduler.py
```
    if request.has_encoder_inputs:
        (encoder_inputs_to_schedule, num_new_tokens,
         new_encoder_compute_budget, external_load_encoder_input
         ) = self._try_schedule_encoder_inputs(
             request, request.num_computed_tokens, num_new_tokens,
             encoder_compute_budget)
```
请求中命中缓存的多模态数据项，如果命中缓存，则不需要重复计算encoder。
会先在显存中查询是否存在encoder cache，显存中不存在，从connector的存储后端中查询。

4) PD实例将需要加载encoder的多模态数据加进metadata(update_state_after_alloc):
vllm/v1/core/sched/scheduler.py
```
    if external_load_encoder_input:
        for i in external_load_encoder_input:
            self.encoder_cache_manager.allocate(request, i)
            if self.ec_connector is not None:
                self.ec_connector.update_state_after_alloc(request, i)
```

5）PD实例加载encoder cache (start_load_caches):
vllm/v1/worker/gpu_model_runner.py
```
  if self.supports_mm_inputs:
            with self.maybe_get_ec_connector_output(
                    scheduler_output,
                    encoder_cache=self.encoder_cache,
            ) as ec_connector_output:
                # Run the multimodal encoder if any.
                self._execute_mm_encoder(scheduler_output)
                mm_embeds = self._gather_mm_embeddings(scheduler_output)
```
计算kvc之后，加载多模态数据的encoder cache.

**2、函数实现**

<img width="1156" height="967" alt="Image" src="https://github.com/user-attachments/assets/d17ba684-dcb0-4c10-b611-a15bab16016d" />

1）定义MooncakeStoreConfig
读取mooncake store配置文件的方法
```
@dataclass
class MooncakeStoreConfig:
   ...

   @staticmethod
    def from_file(file_path: str) -> "MooncakeStoreConfig":
          ....
```

2)  初始化MooncakestoreConnector:
主要是读取mooncake store配置，以及启动store
```
class ECMooncakestoreConnector(ECConnectorBase):

    def __init__(self, vllm_config: "VllmConfig", role: ECConnectorRole):
        super().__init__(vllm_config=vllm_config, role=role)
        
        from mooncake.store import MooncakeDistributedStore
     
        try:
            self.store = MooncakeDistributedStore()
            config_file_path = os.getenv("MOONCAKE_CONFIG_PATH")
            self.config = MooncakeStoreConfig.from_file(config_file_path)
          
            self.store.setup(
              ...
            )

        except ...
          
        # req_id -> index
        self._mm_datas_need_loads: dict[str, int] = {}
        transfer_config = vllm_config.ec_transfer_config
        logger.debug(transfer_config)
```

3）保存encoder cache
```
 def save_caches(self, encoder_cache, mm_hash, **kwargs) -> None:
        """Start saving the EC cache for each mm_datas from encoder cache

        Args:
            **kwargs: additional arguments for the save operation.
        """
        # Return if it is PD Instance
        if not self.is_producer:
            return
        # TODO: 需要讨论，lmcache分配cpu内存有专门的local_cpu_backend,管理内存生命周期以及其他优化。我们是否需要引入
        memory_obj = encoder_cache[mm_hash].detach().cpu()
        kv_bytes = memory_obj.byte_array
        kv_shape = memory_obj.get_shape()
        kv_dtype = memory_obj.get_dtype()
        memory_format = memory_obj.get_memory_format()
        # TODO: 需要添加一个RemoteMetadata类
        metadata_bytes = RemoteMetadata(
            len(kv_bytes), kv_shape, kv_dtype, memory_format
        ).serialize()
        self.store.put(mm_hash, metadata_bytes, kv_bytes)

        logger.debug("Save cache successful for mm_hash %s", mm_hash)
    
```
4）查询encoder cache是否存在
```
    def has_caches(
            self,
            request: "Request",
    ) -> list[bool]:
        result = []
        for mm_hash in request.mm_hashes:
            result.self.store.is_exist(mm_hash)    
        return result
```

5)  加载encode cache:
```
    def start_load_caches(self, **kwargs) -> None:
        """Start loading the EC cache from the connector buffer to worker
        encoder_cache

        Args:
            **kwargs: additional arguments for the load operation
        """

        # Get the metadata
        metadata: ECConnectorMetadata = self._get_connector_metadata()
        assert isinstance(metadata, ECSharedStorageConnectorMetadata)
        encoder_cache = kwargs.get("encoder_cache")  # returns None if missing
        assert encoder_cache is not None
        if metadata is None:
            logger.warning(("In connector.start_load_caches, ",
                            "but the connector metadata is None"))
            return
        # Load the EC for each mm data
        for mm_data in metadata.mm_datas:
            if mm_data.mm_hash in encoder_cache:
                continue
            buffer = self.store.get(mm_data.mm_hash)
            # TODO: 根据Buffer分配一块内存tmp_tensor
            encoder_cache[mm_data.mm_hash] = tmp_tensor
            logger.debug("Success load encoder cache for hash %s",
                         mm_data.mm_hash)
```
